src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/ActivityExtensions.cs

```diff
@@ -0,0 +1,213 @@
+                            Version = activitySource.Version ?? string.Empty, // NRE throw by proto
+                        },
+                    };
+
+                    otlpLibrarySpans.Spans.AddRange(activitySourceEntry.Value);
+                    libraryList.Add(otlpLibrarySpans);
+                }
+
+                var otlpResources = new OtlpResource.Resource();
+                otlpResources.Attributes.AddRange(
+                    resource.Key.Attributes.Select(ToOtlpAttribute));
+
+                var otlpResourceSpans = new OtlpTrace.ResourceSpans
+                {
+                    Resource = otlpResources,
+                };
+                otlpResourceSpans.InstrumentationLibrarySpans.AddRange(libraryList);
+
+                resourceSpansList.Add(otlpResourceSpans);
+            }
+
+            return resourceSpansList;
+        }
+
+        internal static OtlpTrace.Span ToOtlpSpan(this Activity activity)
+        {
+            // protobuf doesn't understand Span<T> yet: https://github.com/protocolbuffers/protobuf/issues/3431
+            Span<byte> traceIdBytes = stackalloc byte[16];
+            Span<byte> spanIdBytes = stackalloc byte[8];
+
+            activity.TraceId.CopyTo(traceIdBytes);
+            activity.SpanId.CopyTo(spanIdBytes);
+
+            var parentSpanIdString = ByteString.Empty;
+            if (activity.ParentSpanId != default)
+            {
+                Span<byte> parentSpanIdBytes = stackalloc byte[8];
+                activity.ParentSpanId.CopyTo(parentSpanIdBytes);
+                parentSpanIdString = ByteString.CopyFrom(parentSpanIdBytes.ToArray());
+            }
+
+            var startTimeUnixNano = activity.StartTimeUtc.ToUnixTimeNanoseconds();
+            var otlpSpan = new OtlpTrace.Span
+            {
+                Name = activity.OperationName, // TODO: map DisplayName to anything?
+
+                Kind = (OtlpTrace.Span.Types.SpanKind)(activity.Kind + 1), // TODO: there is an offset of 1 on the enum.
+
+                TraceId = ByteString.CopyFrom(traceIdBytes.ToArray()),
+                SpanId = ByteString.CopyFrom(spanIdBytes.ToArray()),
+                ParentSpanId = parentSpanIdString,
+
+                // TODO: Status is still pending, need to pursue OTEL spec change.
+
+                StartTimeUnixNano = (ulong)startTimeUnixNano,
+                EndTimeUnixNano = (ulong)(startTimeUnixNano + activity.Duration.ToNanoseconds()),
+            };
+
+            foreach (var kvp in activity.Tags)
+            {
+                // TODO: attempt to convert to supported types?
+                // TODO: enforce no duplicate keys?
+                // TODO: drop if Value is null?
+                // TODO: reverse?
+                var attrib = new OtlpCommon.AttributeKeyValue { Key = kvp.Key, StringValue = kvp.Value };
+                otlpSpan.Attributes.Add(attrib);
```

Comment:
@reyang @cijothomas @noahfalk I think that this is one of the big things to have a wrapper so far: here OTel defines some behaviors that are quite different. In principle, we could have a processor to enforce the OTel spec here but that takes away the control from authors that want to enforce that spec on their code. The OTel related spec: https://github.com/open-telemetry/opentelemetry-specification/blob/bdf39cb935a6f48badc687daeb78a66dc352761f/specification/trace/api.md#set-attributes

Reply:
This is being addressed in Tarek and Noah. We can live with this now, and hopefully get better solution in next preview of DiagnosticSource.

Reply:
Agree this is a valid reason for shim/wrapper to enforce things like no duplicates etc.